### ISSUE TITLE
keyword 'extends' breaks iterative class constructor 

### ISSUE THREAD

**Issue by sotirosn** [jashkenas/coffeescript#3929]
In Chrome, this works fine.

``` coffeescript
class MyClass
   constructor:-> @test(); yield 3
   test:-> console.log 'test ok'
(new MyClass).next()  # outputs 'test ok' then yields 3
```

However if the class extends from a base class then the constructor does not seem to remain a true generator.

``` coffeescript
class MyBase
class MyClass extends MyBase
   constructor:-> @test(); yield 3;
   test:-> console.log 'test ok'
(new MyClass).next()  # throws 'undefined is not a function', because next is undefined
```

This is odd because MyClass.constructor (which is MyClass::constructor.constructor) outputs 'function GeneratorFunction() { [native code] }' when logged to the console.

`MyClass().next` is also undefined, even though `MyClass()` seems to be a 'suspended' generator.

**Comment by sotirosn** [comment]
Ok so if you extend from a constructor that is not a `GeneratorFunction` you lose the `GeneratorFunction::next` method.  So this doesn't work..
```coffeescript
A = -> @test(); yield 3
A.prototype =
   test:-> console.log 'testing'
```
You need to write it like this:
```coffeescript
A = -> @test(); yield 3
A.prototype =
   test:-> console.log 'testing'
   next: A::next
   throw: A::throw
```
to keep next, and throw in the chain.  This seems wrong to me because if you did:
```
A = -> globalArray.push(this); return {objectID: globalArray.indexOf this}
{objectID} = new A
```
then if you wanted to modify `A` so it could be placed in a promise it would become...
```
A = -> globalArray.push(this); yield sleep(1000); return {objectID: globalArray.indexOf this}
{objectID} = yield from new A
```
`new A` should not be a new A object, it should remain an iterator which may return or yield a new A object.  In other words the `this` object should only be exposed internally to the constructor function.  Then the new `this` will have `A`s prototype chain, and `new A` will keep its GeneratorFunction prototype chain.

Another example
```coffeescript
class LinkedList
    collection:[]

    constructor:(@value)->  
        @array.push this
        @index = array.indexOf this
        yield sleep 1000
   
    next:->
        @array[@index+1]
```
Currently `LinkedList::next` conflicts with `GeneratorFunction::next`.  I think `new LinkedList` should remain an iterator with the last return from `(new LinkedList).next()` being the actual object as in `{value:[LinkedList], done:true}`. Of course this is something Chrome and FireFox would need to change so I guess coffeescript could just modify the extend function to check for the existence of 'next' and 'throw' in the child's prototype chain and pass it along if detected.

Then you could do,
```coffeescript
node = new LinkedList(200).next().value
node.next()
```

**Comment by pjeby** [comment]
First off, if you *want* your instances to be generators, shouldn't you extend Generator?  e.g. this works:

```coffeescript
Generator = (-> yield return)().constructor

class MyClass extends Generator
   constructor:-> @test(); yield 3
   test:-> console.log 'test ok'

console.log (new MyClass).next()  # outputs 'test ok' then yields 3
```

Basically, a generator function returns a new Generator instance, and you want your class to be a subclass of that if you want your instances to be generator objects.

As for your second example, the LinkedList thing, it looks like you are trying to write an asynchronous constructor, but in that event you should just return a promise from the constructor, and there's no reason for the constructor to be a generator.  If you're e.g. using `co`, you might write this:

```coffeescript
class LinkedList
    collection:[]

    constructor: co.wrap (@value)->  
        @array.push this
        @index = array.indexOf this
        yield sleep 1000
        return this

    next:->
        @array[@index+1]
```

And then you could use:

```coffeescript
node = yield new LinkedList(200)
node = node.next()  # LinkedList::next
```

To obtain the desired `LinkedList` instance after the sleep was finished.  Similar principles would apply with other generator wrapping libraries.

Short version: ISTM there is no language issue here: you just need to decide whether you want your instances to be generators or not, and return the right thing from your constructor.  (Either by subclassing Generator so your constructor returns a generator, or by writing a constructor that returns a promise or generator or other deferrable that ends by returning `this`.)

**Comment by sotirosn** [comment]
Just because a derived class must be constructed in a series of steps doesn't mean a base class has to as well.  For example I have a base tab view that just basically makes a div with a tab.
```coffeescript
class TabView
   constructor:(name)->
       @element = document.createElement 'div'
       @tab = tabPane.createTab name, this

   close:->
       # stuff to close the tab

   focus:->
      # stuff to focus the tab, etc.
```
Some of these tabs do not necessarily need to be generator constructors, if they contain content already client-side.  But some views do need to load stuff from the server.
```coffeescript
class Editor extends TabView
    constructor:(name, url)->
        super name
        data = yield from http.post url
        # then stuff to load data into the @element
```       
Right now the easiest thing to do for me is to just use a regular constructor and then call an init or load generator method to do the iterative work.  I have seen promises but have come up with a wrapper free solution based on them..

```coffeescript
start = (routine, callback, error, data)->
    {value, done} = if error then routine.throw error else routine.next data
    unless done then value (error, data)->
        start routine, callback, error, data
    else callback value

sleep = (duration)->
    (callback)-> setTimeout (-> callback null), duration

coroutine = (message)->
    yield sleep 500
    console.log message
    yield sleep 500
    message + ' from coroutine'

start do ->
    response = yield (callback)->
        button.onclick = -> callback null, 'Click!'
    console.log response  # 'Click!'

    yield sleep 1000

    response = yield from coroutine 'Hello World!'
    console.log response # 'Hello World! from coroutine'

    tab1 = new TabView 'tab1' # new tab, right away
    tab2 = yield from new Editor 'tab2', 'readfile/72345'  # new tab on ajax response, (but doesn't actually work because TabView is not a generator constructor)
```

I actually use a little more elaborate start routine to try/catch for exceptions, but you get the picture.

**Comment by GeoffreyBooth** [comment]
Neither of the two initial examples compile in CS2. Class constructors cannot be generators in ES2015.
